qemu/tests/cpu_topology_details_test.py

```diff
@@ -0,0 +1,65 @@
+from provider.cpu_utils import check_if_vm_vcpu_topology_match
+
+
+def run(test, params, env):
+    """
+    Check guest gets correct vcpu num, cpu cores, processors, sockets, siblings
+
+    1) Boot guest with options: -smp n,cores=x,threads=y,sockets=z...
+    2) Check cpu topology is expected as set up
+    3) Check if socket number is expected
+    4) Check if cores per socket is expected
+
+    :param test: QEMU test object.
+    :param params: Dictionary with test parameters.
+    :param env: Dictionary with the test environment.
+    """
+
+    def check(p_name, exp, check_cmd):
+        """
+        Check the cpu property inside guest
+        :param p_name: Property name
+        :param exp: The expect value
+        :param check_cmd: The param to get check command
+        """
+        res = session.cmd_output(check_cmd).strip()
+        if int(res) != int(exp):
+            test.fail('The vcpu %s number inside guest is %s,'
+                      ' while it is set to %s' % (p_name, res, exp))
+
+    vm_name = params['main_vm']
+    os_type = params['os_type']
+    vm = env.get_vm(vm_name)
+    session = vm.wait_for_login()
+    if not check_if_vm_vcpu_topology_match(session, os_type, vm.cpuinfo,
+                                           test, vm.devices):
+        test.fail('CPU topology of guest is incorrect.')
+
+    def check_details(vm_session, vm_params, vm_cpuinfo):
```

Comment:
I would move this definition out of the `run` function for better readability. In that case I assume you'd either have to pass `test` as parameter, too, or you could import the exceptions directly:
```
from avocado.core.exceptions import TestFail
...
def check_details(...):
...
             if len(uni_socket) != vm_cpuinfo.sockets:
                raise TestFail('The number of socket is not expected, expect:%s, '
                                         'actual:%s' % (vm_cpuinfo.sockets, len(uni_socket)))
...
```

Reply:
updated